datafusion/core/src/datasource/physical_plan/parquet/statistics.rs

```diff
@@ -766,10 +769,108 @@ impl<'a> StatisticsConverter<'a> {
     }
 
-    /// Returns a null array of data_type with one element per row group
-    fn make_null_array<I>(&self, data_type: &DataType, metadatas: I) -> ArrayRef
+    /// Extract the minimum values from Data page statistics
+    ///
+    /// In Parquet files, in addition to the Column Chunk level statistics
+    /// (stored for each column for each row group) there are also optional
+    /// statistics stored for each data page, part of the [Parquet Page Index].
+    /// Since a single Column Chunk is stored as one or more pages, page level statistics
+    /// can prune at a finer granularity.
+    ///
+    /// However since they are stored in a separate metadata structure
+    /// ([`Index`]) there is different code to extract them as arrow statistics
+    ///
+    /// Parameters:
+    ///
+    /// * `page_index`: The parquet page index, likely read from
+    /// [`ParquetMetadata::page_index()`]
+    ///
+    /// * row_group_indexes: The indexes of the row groups (indexes in
+    /// `page_index`) to extract the statistics from. This is an interator over `&usize` to
+    /// permit passing in  `&Vec<usize>` or similar
+    ///
+    /// # Return Value
+    ///
+    /// The returned array contains 1 value for each `NativeIndex` in the underlying
+    /// `Index`es, in the same order as they appear in `metadatas`.
+    ///
+    /// For example, if there are two `Index`es in `metadatas`:
+    /// 1. the first having `3` `PageIndex` entries
+    /// 2. the second having `2` `PageIndex` entries
+    ///
+    /// The returned array would have 5 rows
+    ///
+    /// Each value is either
+    /// * the minimum value for the page
+    /// * a null value, if the statistics can not be extracted
+    ///
+    /// Note that a null value does NOT mean the min value was actually
+    /// `null` it means it the requested statistic is unknown
+    ///
+    /// # Errors
+    ///
+    /// Reasons for not being able to extract the statistics include:
+    /// * the column is not present in the parquet file
+    /// * statistics for the pages are not present in the row group
+    /// * the stored statistic value can not be converted to the requested type
+    ///
+    /// # Example
+    /// ```no_run
+    /// tood
+    /// ```
+    pub fn data_page_mins<I>(
+        &self,
+        page_index: &ParquetColumnIndex,
+        row_group_indexes: I,
+    ) -> Result<ArrayRef>
+    where
+        I: IntoIterator<Item = &'a usize>,
+    {
+        let data_type = self.arrow_field.data_type();
+
+        let Some(parquet_column_index) = self.parquet_column_index else {
+            return Ok(self.make_null_array(data_type, row_group_indexes));
+        };
+
+        // iterator over &Index
+        let indexes = row_group_indexes
+            .into_iter()
+            .map(|rg_index| &page_index[*rg_index][parquet_column_index]);
+
+        // Get an iterator of the native index type depending on data type
+        match data_type {
```

Comment:
we need to sort out how to make this look better (perhaps by following the lead of the row group iterators that make special iterators for each underlying parquet statistics datatype and then write the relevant code for converting them all to arrow

Reply:
I tried this approach, which works fine. I'd simply refactor whats mentioned in the comment - and extract data type specific iterators (like we already have for the row groups statistics). What do you think @alamb, instead of using array builder?

```Rust
pub(crate) fn min_page_statistics<'a, I: Iterator<Item = Option<&'a Index>>>(
    data_type: Option<&DataType>,
    iterator: I,
) -> Result<ArrayRef> {
    // Extract this into data type specific iterator e.g. MinInt64PageStatisticsIterator
    let iter = iterator.flat_map(|opt_index| match data_type {
        Some(DataType::Int64) => match opt_index {
            Some(Index::INT64(native_index)) => native_index
                .indexes
                .iter()
                .map(|x| x.min)
                .collect::<Vec<_>>(),
            _ => vec![None],
        },
        // other data_types
        _ => todo!(),
    });

    Ok(Arc::new(Int64Array::from_iter(iter)))
```

Reply:
I think an array builder is likely a better approach.